### PR TITLE
Follow up to python rewrite

### PR DESCRIPTION
**Pull request by HW42**
This is based on PR#6 and addresses a bunch of issues not addressed by the rewrite in python (most of those were already there before).

Things that need to be addressed before I think we can release this:

TODO for me:
 - [x] Update README and comments in example config
 - [x] Address https://github.com/HW42/qubes-app-linux-split-gpg2/blob/872550d2e2b8ec2bd99e915fe38ef06b6ee8b3bf/splitgpg2/__init__.py#L483
 - [x] Partially address https://github.com/HW42/qubes-app-linux-split-gpg2/blob/872550d2e2b8ec2bd99e915fe38ef06b6ee8b3bf/splitgpg2/__init__.py#L651

For the last two I have WIP commits (no yet included here since buggy).

Things I would like to have feedback (@marmarek), and then we need to decide:
 - [x] https://github.com/HW42/qubes-app-linux-split-gpg2/pull/6#discussion_r346668633
 - [x] Currently this software is called in most places "split-gpg2" and sometimes "qubes-split-gpg2", do we want to change this and if yes to which version?
 - [x] Which version number should the new release get (since this is not really Qubes release specific probably 0.2 and then after some testing period go to 1.0)
 - [ ] IIRC @marmarek wanted to automate configuration of split-gpg (especially to help new users). The current config file format is not particularly friendly for this and now would be a good time to change it if we want to change it.
 - [ ] Decide about socket binding (see comment below)

### CODE REVIEW

**Comment by marmarek** [comment]
>     * Currently this software is called in most places "split-gpg2" and sometimes "qubes-split-gpg2", do we want to change this and if yes to which version?

I'd go with bare split-gpg2. And definitely keep it separate from "old" split-gpg.

>     * Which version number should the new release get (since this is not really Qubes release specific probably 0.2 and then after some testing period go to 1.0)

Yes, makes sense.

>  IIRC @marmarek wanted to automate configuration of split-gpg (especially to help new users). The current config file format is not particularly friendly for this and now would be a good time to change it if we want to change it.

For the client part, I'd consume old settings - look for `QUBES_GPG_DOMAIN` in the environment and of it isn't there, load it from `/rw/config/gpg-split-domain`. I'd consider an extra fallback to `@default`, to allow setting the target in the qrexec policy.

For the server part, a config file is fine, but IMO there should be also one setting the defaults (to be usable also as a template for user config). Then it can be loaded with whatever method we'll decide:
 - from `/etc/qubes-rpc/qubes.Gpg` - as it is now, but also load the default first
 - from systemd unit when added (using `EnvironmentFile=` setting) - for https://github.com/QubesOS/qubes-core-qrexec/pull/3 support (this would need also handling multiple requests in the same process).
The default one can be for example `/etc/split-gpg.conf`. And the user one I'd name `~/.config/split-gpg.conf`.

As for the socket activation, the original `.socket` file has `Accept=false` and it isn't trivial to modify existing unit file conditionally (depending on `/run/qubes-service/...`). I just hope it's easier to conditionally replace (using `Condition...=...` entry) the unit file actually serving the requests. If so, we'd need to add support for listening socket - I hope there is a socat option for that...
Oh, and the `gpg-agent.socket` is present in Debian, but not Fedora.

But for now, and/or until the above is checked, we can add a script that conditionally create (symlink?) unit files overriding default `gpg-agent.service` and `gpg-agent.socket` (`with Accept=true`). systemd do have provision for dynamically generated unit files, described in `systemd.generator(7)`. I can take care of it.

Anyway, for the user it should be as simple as:
 - installing the package
 - enabling it in the "services" tab in vm settings
 - point client domain to the right server domain with some simple step (preferably vm settings, but we're not there yet)

**Comment by HW42** [comment]
> I'd go with bare split-gpg2.

Done.

---

Re config:  
I have now implemented defaults and systemd wide config. I didn't migrate to systemd style EnvironmentFile since I like being able to script the config. But if you want we can change it.

---

Re socket binding:

There's the problem that if gpg-agent doesn't respond for some reason (until recently for example if you didn't answered the allow prompt) gpg tries to restart it.

I think we have 3 options:
 1. script as currently (maybe make `.service` as user service?)
 2. systemd socket activation
 3. replace gpg-agent with a wrapper

1 and 2 have the restart problem.
3 is easy to do implement with `dpkg-divert` on Debian but I'm not aware how to move the real gpg-agent cleanly on Fedora.

Since socket activation doesn't have a big advantage (the script is only started if someone enabled the qubes-service) and 3 seems to hacky I stayed with option 1 for now. But I'm fine to change it.

**Comment by HW42** [comment]
Except for the running discussion (and whatever bugs we find in the meantime) I think this is now in a good shape for a release to get some wider testing.

@marmarek: Please have a look at this branch. If you want feel free to squash some of my fixup commits into your rewrite commit (I will rebase after you have done this).

If these points are resolved I want to release 0.2. Which I think we then should upload to unstable and announce on qubes-devel to get a few tests. And later a 1.0 release in the normal repo.

**Comment by pietrushnic** [comment]
Hi @HW42 , @marmarek ,
I would like to take a look at this and if the possible move that forward. Also, I'm thinking about having an introductory presentation about this project at Qubes OS mini-summit 2021.

I have some minor README fixes and some clarification questions, which I will post here. If you know anyone who can help to move this project, at least from a review perspective, please let me know.

**Review by pietrushnic — CHANGES_REQUESTED** [review]

**Comment by pietrushnic** [comment]
@marmarek small typo fixes: https://github.com/HW42/qubes-app-linux-split-gpg2/pull/8

File: README.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+Add it as a component to [qubes-builder](https://github.com/QubesOS/qubes-builder) and built it.
+
+For development you can also build the Debian packet in tree.
+Just run `dpkg-buildpackage -us -uc` at the top level.
+
+## Installation
+
+Install the the deb or the rpm on your TemplateVM(s).
+
+Install the dom0-rpm in dom0 or just create `/etc/qubes-rpc/policy/qubes.Gpg2` by hand.
```
[the remark this answers is not in the document]
Comment (marmarek) [reply README.md:36]: Yes, this should be good enough for initial test, although the number of prompts may be annoying. You may want to add things like `<your client VM> <your gpg-holding VM> allow`

File: README.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+```
+gpg-server-vm$ gpg --gen-key
+```
+
+Now configure which domain the client VM should use as server. Either:
+ - Set the target for `@default` in qubes.Gpg2. or
+ - Write `SPLIT_GPG2_SERVER_DOMAIN=<gpg-server>` into `~user/.config/.split-gpg2-rc`.
+
+Enable the `split-gpg2-client` service in the client domain either via the GUI or via `qvm-service gpg-client-vm split-gpg2-client on`.
+Restart the client domain.
```
[the remark this answers is not in the document]
Comment (marmarek) [reply README.md:56]: You can add a service `split-gpg2-client` in the service tab by manually entering it. Having this better integrated is one of the missing points of this work...

File: README.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
+# split-gpg2
+
+## About split-gpg2
+
+This software allows you to separate the handling of private key material from the rest of GnuPG's processing.
+This is similar to how smartcards work, except that in this case the handling of the private key is not put into some small microcontroller but into another Qubes domain.
```
[the remark this answers is not in the document]
Comment (marmarek) [reply README.md:6]: Generally [for documentation we prefer one sentence per line](https://www.qubes-os.org/doc/doc-guidelines/#markdown-conventions), which makes it easier to translate, but also easier to review changes (no unneeded churn by re-wrapping).

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (pietrushnic) [reply README.md:36]: Added something like that.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (pietrushnic) [reply README.md:56]: Ok, tested that and explained it better in readme fixes PR.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (pietrushnic) [reply README.md:6]: Ok, I tried to improve my PR according to rules.

**Comment by HW42** [comment]
Merged to master excluding the WIP version commit.

**Pull request closed without merging**